// scratch-vm/src/blockly/runtime/CodeProfiler.js

class CodeProfiler
{
	constructor(){
		this.infoDict = {};
		this.timestamp = 0;
	}
	
	reset(){
		for(var info of this.infoDict){
			info[0] = info[1] = 0;
		}
	}
	
	begin(key){
		if(!(key in this.infoDict)){
			infoDict[key] = [0, 0];
		}
		this.timestamp = performance.now();
	}
	
	end(key){
		var info = this.infoDict[key];
		info[0] += performance.now() - this.timestamp;
		info[1] += 1;
	}
	
	print(){
		var infoList = [];
		var info;
		var total = 0;
		for(var key in this.infoDict){
			info = this.infoDict[key];
			if(info[0] <= 0){
				continue;
			}
			total += info[0];
			var value = info[0] * 100;
			infoList.push([key, value, value / info[1]]);
		}
		if(total <= 0){
			return;
		}
		console.log("/**Profile**/", total);
		infoList.sort(_sortInfo);
		for(info of infoList){
			console.log(Math.round(info[1] / total) + "%\t\t" + Math.round(info[2] / total) + "%\t\t" + info[0]);
		}
	}
}

function _sortInfo(a, b)
{
	if (a[1] > b[1]) return -1;
	if (a[1] < b[1]) return  1;
	if (a[2] > b[2]) return -1;
	if (a[2] < b[2]) return  1;
	return 0;
}

module.exports = CodeProfiler;
